torch/package/module_environment.py

```diff
@@ -0,0 +1,219 @@
+        self.modules = sys.modules
+
+    def import_module(self, module_name: str):
+        return importlib.import_module(module_name)
+
+
+DefaultImporter = _DefaultImporter()
+
+
+class ModuleEnv:
+    """An environment that manages multiple module importers with overlapping names.
+
+    By default, you can figure out what module an object belongs by checking
+    __module__ and importing the result using __import__ or importlib.import_module.
+
+    torch.package introduces module importers other than the default one.
+    Each PackageImporter introduces a new namespace. Potentially a single
+    name (e.g. 'foo.bar') is present in multiple namespaces.
+
+    ModuleEnv provides a structured way to represent a collection of
+    importers and ensuring that we are retrieving the correct modules for a
+    given name.
+
+    It supports two main operations:
+        get_name: object -> (parent module name, name of obj within module)
+        import_module: module_name -> module
+
+    The guarantee is that following round-trip will succeed or throw a ModuleEnvError.
+        module_name, obj_name = env.get_name(obj)
+        module = env.import_module(module_name)
+        obj2 = getattr(module, obj_name)
+        assert obj1 is obj2
+    """
+
+    def __init__(self, importers: Optional[List[BaseImporter]] = None):
+        if importers is None:
+            importers = [DefaultImporter]
+        self._importers = importers
+
+    def import_module(self, module_name: str) -> ModuleType:
+        """Import 'module_name' from this environment.
+
+        Importers will be tried in order of `self.importers`, and the first match will be taken.
+        """
+        last_err = None
+        for importer in self._importers:
+            if not isinstance(importer, BaseImporter):
+                raise TypeError(
+                    f"{importer} is not a BaseImporter. "
+                    "All importers in ModuleEnv must inherit from BaseImporter."
+                )
+            try:
+                return importer.import_module(module_name)
+            except ModuleNotFoundError as err:
+                last_err = err
+
+        if last_err is not None:
+            raise last_err
+        else:
+            raise ModuleNotFoundError(module_name)
+
+    def get_name(
+        self, obj: Any, name: Optional[str] = None, check_obj: bool = True
+    ) -> Tuple[str, str]:
+        """Given an object, return a name that can be used to retrieve the
+        object from this environment.
+
+        Args:
+            obj: An object to get the the module-environment-relative name for.
+            name: If set, use this name instead of looking up __name__ or __qualname__ on `obj`.
+                This is only here to match how Pickler handles __reduce__ functions that return a string,
+                don't use otherwise.
+            check_obj: If true, check that retrieving the returned name produces the same object as `obj`.
```

Comment:
this flag feels weird. This should either always return None or always throw an exception if it cannot find the object.

I feel like the problem is that the error reporting for the mismatch is still happening in the pickler,  rather than just reporting the right error here: e.g. I found the object but it is not the same object directly in the exception.

Reply:
Yeah, I was sort of struggling to not rewrite the pickler code too much. But maybe for the sake of this API not being silly it makes sense to be more intrusive (since we're editing the pickler code anyway…)